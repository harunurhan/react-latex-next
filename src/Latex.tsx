import * as React from 'react'

import renderLatex from './renderLatex';
import { Delimiter } from './types';
import './Latex.css'

export interface LatexProps {
  children: string;
  delimiters?: Delimiter[];
  strict?: boolean;
  macros?: object
}

export default class Latex extends React.Component<LatexProps> {
  static defaultProps: Partial<LatexProps> = {
    delimiters: [
      { left: '$$', right: '$$', display: true },
      { left: '\\(', right: '\\)', display: false },
      { left: '$', right: '$', display: false },
      { left: '\\[', right: '\\]', display: true },
    ],
    strict: false
  };

  render() {
    const { children, delimiters, strict, macros } = this.props
    const renderedLatex = renderLatex(children, delimiters!, strict!, macros);
    return (
      <span className="__Latex__" dangerouslySetInnerHTML={{ __html: renderedLatex }} />
    )
  }
}
