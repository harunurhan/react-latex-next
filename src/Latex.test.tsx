import * as React from 'react';
import { shallow } from 'enzyme';
import Latex from './Latex';

describe('Latex', () => {
  it('renders text that has multiple LaTeX formulas with $ delimiter', () => {
    const latex = 'three processes $e^+e^-$, gluon-gluon and $\\gamma\\gamma \\to W t\\bar b$.';
    const wrapper = shallow(<Latex>{latex}</Latex>);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders text that has multiple LaTeX formulas with $$ delimiter', () => {
    const latex = 'three processes $$e^+e^-$$, gluon-gluon and $$\\gamma\\gamma \\to W t\\bar b$$.';
    const wrapper = shallow(<Latex>{latex}</Latex>);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders text that has a LaTeX formula with custom delimiter', () => {
    const latex = 'three processes <math e^+e^- math>';
    const delimiters = [
      { left: '<math ', right: ' math>', display: true },
    ]
    const wrapper = shallow(<Latex delimiters={delimiters}>{latex}</Latex>);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders raw LaTeX incase of error by default', () => {
    const latex = 'Broken formulate: $\to W t\bar b$.';
    const wrapper = shallow(<Latex>{latex}</Latex>);
    expect(wrapper).toMatchSnapshot();
  });

  it('throws error incase of error', () => {
    const latex = 'Broken formulate: $\to W t\bar b$.';
    const brokenRender = () => shallow(<Latex strict>{latex}</Latex>);
    expect(brokenRender).toThrowError();
  });

  it("renders correctly sequences of $..$", () => {
    const latex = "$hello$$world$$boo$";
    const wrapper = shallow(<Latex>{latex}</Latex>);
    expect(wrapper).toMatchSnapshot();
  });

  it("renders an expression with macros", () => {
    const latex = "$\\R$";
    const wrapper = shallow(<Latex macros={{"\\R": "\\mathbb{R}"}}>{latex}</Latex>);
    expect(wrapper).toMatchSnapshot();
  })
})
