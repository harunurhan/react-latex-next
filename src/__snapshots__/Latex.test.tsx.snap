// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Latex renders an expression with macros 1`] = `
<span
  className="__Latex__"
  dangerouslySetInnerHTML={
    Object {
      "__html": "<span class=\\"katex\\"><span class=\\"katex-mathml\\"><math xmlns=\\"http://www.w3.org/1998/Math/MathML\\"><semantics><mrow><mi mathvariant=\\"double-struck\\">R</mi></mrow><annotation encoding=\\"application/x-tex\\">\\\\R</annotation></semantics></math></span><span class=\\"katex-html\\" aria-hidden=\\"true\\"><span class=\\"base\\"><span class=\\"strut\\" style=\\"height:0.68889em;vertical-align:0em;\\"></span><span class=\\"mord mathbb\\">R</span></span></span></span>",
    }
  }
/>
`;

exports[`Latex renders correctly sequences of $..$ 1`] = `
<span
  className="__Latex__"
  dangerouslySetInnerHTML={
    Object {
      "__html": "<span class=\\"katex\\"><span class=\\"katex-mathml\\"><math xmlns=\\"http://www.w3.org/1998/Math/MathML\\"><semantics><mrow><mi>h</mi><mi>e</mi><mi>l</mi><mi>l</mi><mi>o</mi></mrow><annotation encoding=\\"application/x-tex\\">hello</annotation></semantics></math></span><span class=\\"katex-html\\" aria-hidden=\\"true\\"><span class=\\"base\\"><span class=\\"strut\\" style=\\"height:0.69444em;vertical-align:0em;\\"></span><span class=\\"mord mathnormal\\">h</span><span class=\\"mord mathnormal\\">e</span><span class=\\"mord mathnormal\\" style=\\"margin-right:0.01968em;\\">ll</span><span class=\\"mord mathnormal\\">o</span></span></span></span><span class=\\"katex\\"><span class=\\"katex-mathml\\"><math xmlns=\\"http://www.w3.org/1998/Math/MathML\\"><semantics><mrow><mi>w</mi><mi>o</mi><mi>r</mi><mi>l</mi><mi>d</mi></mrow><annotation encoding=\\"application/x-tex\\">world</annotation></semantics></math></span><span class=\\"katex-html\\" aria-hidden=\\"true\\"><span class=\\"base\\"><span class=\\"strut\\" style=\\"height:0.69444em;vertical-align:0em;\\"></span><span class=\\"mord mathnormal\\" style=\\"margin-right:0.02691em;\\">w</span><span class=\\"mord mathnormal\\" style=\\"margin-right:0.02778em;\\">or</span><span class=\\"mord mathnormal\\" style=\\"margin-right:0.01968em;\\">l</span><span class=\\"mord mathnormal\\">d</span></span></span></span><span class=\\"katex\\"><span class=\\"katex-mathml\\"><math xmlns=\\"http://www.w3.org/1998/Math/MathML\\"><semantics><mrow><mi>b</mi><mi>o</mi><mi>o</mi></mrow><annotation encoding=\\"application/x-tex\\">boo</annotation></semantics></math></span><span class=\\"katex-html\\" aria-hidden=\\"true\\"><span class=\\"base\\"><span class=\\"strut\\" style=\\"height:0.69444em;vertical-align:0em;\\"></span><span class=\\"mord mathnormal\\">b</span><span class=\\"mord mathnormal\\">oo</span></span></span></span>",
    }
  }
/>
`;

exports[`Latex renders raw LaTeX incase of error by default 1`] = `
<span
  className="__Latex__"
  dangerouslySetInnerHTML={
    Object {
      "__html": "Broken formulate: 	o W tar b.",
    }
  }
/>
`;

exports[`Latex renders text that has a LaTeX formula with custom delimiter 1`] = `
<span
  className="__Latex__"
  dangerouslySetInnerHTML={
    Object {
      "__html": "three processes <span class=\\"katex-display\\"><span class=\\"katex\\"><span class=\\"katex-mathml\\"><math xmlns=\\"http://www.w3.org/1998/Math/MathML\\" display=\\"block\\"><semantics><mrow><msup><mi>e</mi><mo>+</mo></msup><msup><mi>e</mi><mo>−</mo></msup></mrow><annotation encoding=\\"application/x-tex\\">e^+e^-</annotation></semantics></math></span><span class=\\"katex-html\\" aria-hidden=\\"true\\"><span class=\\"base\\"><span class=\\"strut\\" style=\\"height:0.821331em;vertical-align:0em;\\"></span><span class=\\"mord\\"><span class=\\"mord mathnormal\\">e</span><span class=\\"msupsub\\"><span class=\\"vlist-t\\"><span class=\\"vlist-r\\"><span class=\\"vlist\\" style=\\"height:0.821331em;\\"><span style=\\"top:-3.113em;margin-right:0.05em;\\"><span class=\\"pstrut\\" style=\\"height:2.7em;\\"></span><span class=\\"sizing reset-size6 size3 mtight\\"><span class=\\"mbin mtight\\">+</span></span></span></span></span></span></span></span><span class=\\"mord\\"><span class=\\"mord mathnormal\\">e</span><span class=\\"msupsub\\"><span class=\\"vlist-t\\"><span class=\\"vlist-r\\"><span class=\\"vlist\\" style=\\"height:0.821331em;\\"><span style=\\"top:-3.113em;margin-right:0.05em;\\"><span class=\\"pstrut\\" style=\\"height:2.7em;\\"></span><span class=\\"sizing reset-size6 size3 mtight\\"><span class=\\"mbin mtight\\">−</span></span></span></span></span></span></span></span></span></span></span></span>",
    }
  }
/>
`;

exports[`Latex renders text that has multiple LaTeX formulas with $ delimiter 1`] = `
<span
  className="__Latex__"
  dangerouslySetInnerHTML={
    Object {
      "__html": "three processes <span class=\\"katex\\"><span class=\\"katex-mathml\\"><math xmlns=\\"http://www.w3.org/1998/Math/MathML\\"><semantics><mrow><msup><mi>e</mi><mo>+</mo></msup><msup><mi>e</mi><mo>−</mo></msup></mrow><annotation encoding=\\"application/x-tex\\">e^+e^-</annotation></semantics></math></span><span class=\\"katex-html\\" aria-hidden=\\"true\\"><span class=\\"base\\"><span class=\\"strut\\" style=\\"height:0.771331em;vertical-align:0em;\\"></span><span class=\\"mord\\"><span class=\\"mord mathnormal\\">e</span><span class=\\"msupsub\\"><span class=\\"vlist-t\\"><span class=\\"vlist-r\\"><span class=\\"vlist\\" style=\\"height:0.771331em;\\"><span style=\\"top:-3.063em;margin-right:0.05em;\\"><span class=\\"pstrut\\" style=\\"height:2.7em;\\"></span><span class=\\"sizing reset-size6 size3 mtight\\"><span class=\\"mbin mtight\\">+</span></span></span></span></span></span></span></span><span class=\\"mord\\"><span class=\\"mord mathnormal\\">e</span><span class=\\"msupsub\\"><span class=\\"vlist-t\\"><span class=\\"vlist-r\\"><span class=\\"vlist\\" style=\\"height:0.771331em;\\"><span style=\\"top:-3.063em;margin-right:0.05em;\\"><span class=\\"pstrut\\" style=\\"height:2.7em;\\"></span><span class=\\"sizing reset-size6 size3 mtight\\"><span class=\\"mbin mtight\\">−</span></span></span></span></span></span></span></span></span></span></span>, gluon-gluon and <span class=\\"katex\\"><span class=\\"katex-mathml\\"><math xmlns=\\"http://www.w3.org/1998/Math/MathML\\"><semantics><mrow><mi>γ</mi><mi>γ</mi><mo>→</mo><mi>W</mi><mi>t</mi><mover accent=\\"true\\"><mi>b</mi><mo>ˉ</mo></mover></mrow><annotation encoding=\\"application/x-tex\\">\\\\gamma\\\\gamma \\\\to W t\\\\bar b</annotation></semantics></math></span><span class=\\"katex-html\\" aria-hidden=\\"true\\"><span class=\\"base\\"><span class=\\"strut\\" style=\\"height:0.625em;vertical-align:-0.19444em;\\"></span><span class=\\"mord mathnormal\\" style=\\"margin-right:0.05556em;\\">γγ</span><span class=\\"mspace\\" style=\\"margin-right:0.2777777777777778em;\\"></span><span class=\\"mrel\\">→</span><span class=\\"mspace\\" style=\\"margin-right:0.2777777777777778em;\\"></span></span><span class=\\"base\\"><span class=\\"strut\\" style=\\"height:0.8312199999999998em;vertical-align:0em;\\"></span><span class=\\"mord mathnormal\\" style=\\"margin-right:0.13889em;\\">W</span><span class=\\"mord mathnormal\\">t</span><span class=\\"mord accent\\"><span class=\\"vlist-t\\"><span class=\\"vlist-r\\"><span class=\\"vlist\\" style=\\"height:0.8312199999999998em;\\"><span style=\\"top:-3em;\\"><span class=\\"pstrut\\" style=\\"height:3em;\\"></span><span class=\\"mord mathnormal\\">b</span></span><span style=\\"top:-3.26344em;\\"><span class=\\"pstrut\\" style=\\"height:3em;\\"></span><span class=\\"accent-body\\" style=\\"left:-0.25em;\\"><span class=\\"mord\\">ˉ</span></span></span></span></span></span></span></span></span></span>.",
    }
  }
/>
`;

exports[`Latex renders text that has multiple LaTeX formulas with $$ delimiter 1`] = `
<span
  className="__Latex__"
  dangerouslySetInnerHTML={
    Object {
      "__html": "three processes <span class=\\"katex-display\\"><span class=\\"katex\\"><span class=\\"katex-mathml\\"><math xmlns=\\"http://www.w3.org/1998/Math/MathML\\" display=\\"block\\"><semantics><mrow><msup><mi>e</mi><mo>+</mo></msup><msup><mi>e</mi><mo>−</mo></msup></mrow><annotation encoding=\\"application/x-tex\\">e^+e^-</annotation></semantics></math></span><span class=\\"katex-html\\" aria-hidden=\\"true\\"><span class=\\"base\\"><span class=\\"strut\\" style=\\"height:0.821331em;vertical-align:0em;\\"></span><span class=\\"mord\\"><span class=\\"mord mathnormal\\">e</span><span class=\\"msupsub\\"><span class=\\"vlist-t\\"><span class=\\"vlist-r\\"><span class=\\"vlist\\" style=\\"height:0.821331em;\\"><span style=\\"top:-3.113em;margin-right:0.05em;\\"><span class=\\"pstrut\\" style=\\"height:2.7em;\\"></span><span class=\\"sizing reset-size6 size3 mtight\\"><span class=\\"mbin mtight\\">+</span></span></span></span></span></span></span></span><span class=\\"mord\\"><span class=\\"mord mathnormal\\">e</span><span class=\\"msupsub\\"><span class=\\"vlist-t\\"><span class=\\"vlist-r\\"><span class=\\"vlist\\" style=\\"height:0.821331em;\\"><span style=\\"top:-3.113em;margin-right:0.05em;\\"><span class=\\"pstrut\\" style=\\"height:2.7em;\\"></span><span class=\\"sizing reset-size6 size3 mtight\\"><span class=\\"mbin mtight\\">−</span></span></span></span></span></span></span></span></span></span></span></span>, gluon-gluon and <span class=\\"katex-display\\"><span class=\\"katex\\"><span class=\\"katex-mathml\\"><math xmlns=\\"http://www.w3.org/1998/Math/MathML\\" display=\\"block\\"><semantics><mrow><mi>γ</mi><mi>γ</mi><mo>→</mo><mi>W</mi><mi>t</mi><mover accent=\\"true\\"><mi>b</mi><mo>ˉ</mo></mover></mrow><annotation encoding=\\"application/x-tex\\">\\\\gamma\\\\gamma \\\\to W t\\\\bar b</annotation></semantics></math></span><span class=\\"katex-html\\" aria-hidden=\\"true\\"><span class=\\"base\\"><span class=\\"strut\\" style=\\"height:0.625em;vertical-align:-0.19444em;\\"></span><span class=\\"mord mathnormal\\" style=\\"margin-right:0.05556em;\\">γγ</span><span class=\\"mspace\\" style=\\"margin-right:0.2777777777777778em;\\"></span><span class=\\"mrel\\">→</span><span class=\\"mspace\\" style=\\"margin-right:0.2777777777777778em;\\"></span></span><span class=\\"base\\"><span class=\\"strut\\" style=\\"height:0.8312199999999998em;vertical-align:0em;\\"></span><span class=\\"mord mathnormal\\" style=\\"margin-right:0.13889em;\\">W</span><span class=\\"mord mathnormal\\">t</span><span class=\\"mord accent\\"><span class=\\"vlist-t\\"><span class=\\"vlist-r\\"><span class=\\"vlist\\" style=\\"height:0.8312199999999998em;\\"><span style=\\"top:-3em;\\"><span class=\\"pstrut\\" style=\\"height:3em;\\"></span><span class=\\"mord mathnormal\\">b</span></span><span style=\\"top:-3.26344em;\\"><span class=\\"pstrut\\" style=\\"height:3em;\\"></span><span class=\\"accent-body\\" style=\\"left:-0.25em;\\"><span class=\\"mord\\">ˉ</span></span></span></span></span></span></span></span></span></span></span>.",
    }
  }
/>
`;
